src/OpenApi/src/Extensions/JsonNodeSchemaExtensions.cs

```diff
@@ -327,10 +327,19 @@ internal static void ApplyParameterInfo(this JsonNode schema, ApiParameterDescri
     /// <param name="createSchemaReferenceId">A delegate that generates the reference ID to create for a type.</param>
     internal static void ApplyPolymorphismOptions(this JsonNode schema, JsonSchemaExporterContext context, Func<JsonTypeInfo, string?> createSchemaReferenceId)
     {
-        // The `context.Path.Length == 0` check is used to ensure that we only apply the polymorphism options
+        // The `context.BaseTypeInfo == null` check is used to ensure that we only apply the polymorphism options
         // to the top-level schema and not to any nested schemas that are generated.
-        if (context.TypeInfo.PolymorphismOptions is { } polymorphismOptions && context.Path.Length == 0)
+        if (context.TypeInfo.PolymorphismOptions is { } polymorphismOptions && context.BaseTypeInfo == null)
         {
+            // System.Text.Json supports serializing to a non-abstract base class if no discriminator is provided.
+            // OpenAPI requires that all polymorphic sub-schemas have an associated discriminator. To maintain accuracy,
+            // with the actual implementation and comply with the limitations of the OpenAPI schema we do not emit
+            // a `discriminator` property for types that are represented this way.
+            if (!context.TypeInfo.Type.IsAbstract &&
+                !polymorphismOptions.DerivedTypes.Any(type => type.DerivedType == context.TypeInfo.Type))
+            {
+                return;
```

Comment:
I'm reading this early return as "don't apply the polymorphism options", but I don't fully understand the consequences of that.  There's some explanation in your PR description, but it might be helpful to incorporate some of that into the code.

I think maybe "ApplyPolymorphismOptions" means "include discriminator properties" and the early return is a way of saying "don't include discriminator properties", which will have the effect of emitting "any" (which I think is an undiscriminated union, rather than a wildcard).  Is that right?

Reply:
> I think maybe "ApplyPolymorphismOptions" means "include discriminator properties" and the early return is a way of saying "don't include discriminator properties", which will have the effect of emitting "any" (which I think is an undiscriminated union, rather than a wildcard). Is that right?

Yep. I've updated the name of the method to be a little be clearer about how the polymorphism options are mapped to the OpenAPI document.